### PR TITLE
Debugger: Support setting line number and begin of line offset when calling parser

### PR DESCRIPTION
**Pull request by jfehrle**
c518c206ae7a1e7756170e32c2061470dbe88f22 allowed the caller to set the buffer offset when initializing the parser.  This change allows setting the line number and the beginning of line offset.

CoqIDE invokes the parser separately for each statement.  The earlier change made it possible to return offsets relative to the beginning of the buffer, which was sufficient for CoqIDE to highlight the code being executed.  The stack panel in the debugger shows the line number.  This change ensures that Coq returns correct line numbers in the Loc.t.

I extracted this code from #14644 to make the review of this piece easy.

### CODE REVIEW

File: gramlib/plexing.mli
Comment (ejgallego) [inline gramlib/plexing.mli:35]: Please use named parameters instead of a tuple.

File: parsing/cLexer.ml
Comment (ejgallego) [inline parsing/cLexer.ml:395]: Similarly here, use a record.

**Review by ejgallego — CHANGES_REQUESTED** [review]
The change is fine, however the API is dubious [coming from the previous commit]

The offset should be set only once per lexer initalization, and thus, the `set_loc_offset` should be removed from the public API and instead the offset should be part of init.

Moreover, that loc should be then set as the initial loc and avoid the adjustment.

File: gramlib/plexing.mli
Comment (jfehrle) [reply gramlib/plexing.mli:35]: I tried to do that a month ago but wasn't able to get it compile.  IIRC it was a combination of visibility of the type and circular linking.

**Comment by jfehrle** [comment]
> The offset should be set only once per lexer initalization, and thus, the `set_loc_offset` should be removed from the public API and instead the offset should be part of init.

That was the first thing I tried.  IIRC that broke other CoqIDE functionality that I had not otherwise touched.  It didn't look like a simple fix.

File: gramlib/plexing.mli
Comment (ejgallego) [reply gramlib/plexing.mli:35]: I am just suggesting to do `set_loc_offset : bl:int -> el:int -> ....`

Comment (jfehrle) [reply gramlib/plexing.mli:35]: Indeed, I misinterpreted your comment, thinking about records instead.

File: parsing/cLexer.ml
Comment (jfehrle) [reply parsing/cLexer.ml:395]: Although you mention records here.

**Comment by jfehrle** [comment]
Part of why I did it this way is that `tok_func` is buried inside `cLexer.MakeLexer`. `MakeLexer` is only used to create module `Lexer` in the same file.  I didn't see an easy way to pass the loc into `tok_func` from `Grammar.Parsable.make`.  I recall spending a couple hours trying various alternatives without success.  OCaml's module/functor documentation is not the greatest.  If you know how to make this work, preferably by pushing another commit, that would be great.

The other part of it is that starting the lexer with an offset other than 0 caused IIRC script highlighting for error messages to appear in the wrong place.  I only discovered that by chance.  I was concerned that there might be other code paths that a 
non-zero initial offset would break with no great way for me to discover them.  So I made a more conservative change that couldn't break anything else in CoqIDE.

> Actually, can you remind me why this and #14220 are needed in the first place?

#14220 is needed so the CoqIDE can highlight the code at the stopping point/breakpoint.  To do this, the debugger must return an offset relative to the beginning of the buffer/script file rather than relative to the beginning of the statement.  This ultimately comes from the AST, the output of the parser.

This additional change is to make the line number available so it can be shown to the user in the stack frame.

**Comment by ejgallego** [comment]
Looks better now, thanks! Still a bit puzzled about what's going on in `next_token` so I'll ping the rest of maintainers for feedback before merge.

**Review by ejgallego — APPROVED** [review]

**Pull request merged**
